src/BaseControllerV2.test.ts

```diff
@@ -511,4 +571,149 @@ describe('getPersistentState', () => {
+          messenger: messagingSystem,
+          metadata: visitorControllerStateMetadata,
+          name: 'VisitorController',
+          state: { visitors: [] },
+        });
+        messagingSystem.registerActionHandler('VisitorController:clear', this.clear);
+      }
+
+      clear = () => {
+        this.update(() => {
+          return { visitors: [] };
+        });
+      };
+
+      addVisitor(visitor: string) {
+        this.update(({ visitors }) => {
+          return { visitors: [...visitors, visitor] };
+        });
+      }
+
+      destroy() {
+        super.destroy();
+      }
+    }
+
+    type VisitorOverflowControllerState = {
+      maxVisitors: number;
+    };
+    type VisitorOverflowControllerAction = {
+      type: `VisitorOverflowController:updateMax`;
+      handler: (max: number) => void;
+    };
+    type VisitorOverflowControllerEvent = {
+      type: `VisitorOverflowController:stateChange`;
+      payload: [VisitorOverflowControllerState, Patch[]];
+    };
+
+    const visitorOverflowControllerMetadata = {
+      maxVisitors: {
+        persist: false,
+        anonymous: true,
+      },
+    };
+
+    class VisitorOverflowController extends BaseController<
+      'VisitorOverflowController',
+      VisitorOverflowControllerState
+    > {
+      constructor(
+        messagingSystem: RestrictedControllerMessenger<
+          'VisitorOverflowController',
+          VisitorControllerAction | VisitorOverflowControllerAction,
+          VisitorControllerEvent | VisitorOverflowControllerEvent,
+          'VisitorController:clear',
+          'VisitorController:stateChange'
+        >,
+      ) {
+        super({
+          messenger: messagingSystem,
+          metadata: visitorOverflowControllerMetadata,
+          name: 'VisitorOverflowController',
+          state: { maxVisitors: 5 },
+        });
+        messagingSystem.registerActionHandler('VisitorOverflowController:updateMax', this.updateMax);
+        messagingSystem.subscribe('VisitorController:stateChange', this.onVisit);
+      }
+
+      onVisit = ({ visitors }: VisitorControllerState) => {
+        if (visitors.length > this.state.maxVisitors) {
+          this.messagingSystem.call('VisitorController:clear');
+        }
+      };
+
+      updateMax = (max: number) => {
+        this.update(() => {
+          return { maxVisitors: max };
+        });
+      };
+
+      destroy() {
+        super.destroy();
+      }
+    }
+
+    it('should allow messaging between controllers', () => {
+      const controllerMessenger = new ControllerMessenger<
+        VisitorControllerAction | VisitorOverflowControllerAction,
+        VisitorControllerEvent | VisitorOverflowControllerEvent
+      >();
+      const visitorControllerMessenger = controllerMessenger.getRestricted({
+        name: 'VisitorController',
+        allowedActions: [],
+        allowedEvents: [],
+      });
+      const visitorController = new VisitorController(visitorControllerMessenger);
```

Comment:
I thought we'd moved to an options bag for this type of pattern, is that optional?

Reply:
The base controller and the messaging system both use the options bag pattern. There are no constraints upon controller constructors - I don't think it's possible for us to constrain that :thinking:.

This was just a mock controller, which I kept as simple as possible to reduce the size of this implementation (which ended up being rather large anyway, as you can see). I was also thinking that if we did have a controller with zero options, it would be OK for it to accept the messaging system directly without an options bag. I could update it to use an options bag parameter anyway though, for consistency.